lib/utilities/retrieveData.js

```diff
@@ -1,31 +1,73 @@
-export default function retrieveData(locations, dataRetrievalFn,
-    response, logger, errorHandlerFn) {
-    if (locations.length === 0) {
-        return response.end();
-    }
-    if (errorHandlerFn === undefined) {
-        // eslint-disable-next-line
-        errorHandlerFn = () => { response.connection.destroy(); };
-    }
-    const current = locations.shift();
-    return dataRetrievalFn(current, logger,
-        (err, readable) => {
+const { eachSeries } = require('async');
+const responseErr = new Error();
+responseErr.code = 'ResponseError';
+responseErr.message = 'response errored out before readable finished';
+
+export default function retrieveData(locations, retrieveDataFn, response, log) {
+    let responseDestroyed = false;
+    const _destroyResponse = () => {
+        // destroys the socket if available
+        response.destroy();
+        responseDestroyed = true;
+    };
+    response.on('close', () => {
```

Comment:
why 'on' here but 'once' in sproxydclient and below?

Reply:
Just being careful here to avoid EventEmitter leak since it's in a loop.